File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/PostTableConnection.java
package com.emogoth.android.phone.mimi.db;


import com.emogoth.android.phone.mimi.db.model.PostModel;
import com.mimireader.chanlib.models.ChanPost;
import com.mimireader.chanlib.models.ChanThread;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Single;
import io.reactivex.functions.Function;

public class PostTableConnection {
    public static final String LOG_TAG = PostTableConnection.class.getSimpleName();

    public static Single<List<PostModel>> fetchThread(long threadId) {
        return DatabaseUtils.fetchTable(PostModel.class, PostModel.TABLE_NAME, "post_id", PostModel.KEY_THREAD_ID + "=?", String.valueOf(threadId));
    }

    public static Flowable<List<PostModel>> watchThread(long threadId) {
        return DatabaseUtils.observeTable(PostModel.class, PostModel.TABLE_NAME, "post_id", PostModel.KEY_THREAD_ID + "=?", String.valueOf(threadId));
    }

    public static Function<List<PostModel>, ChanThread> mapDbPostsToChanThread(final String boardName, final long threadId) {
        return postModels -> convertDbPostsToChanThread(boardName, threadId, postModels);
    }

    public static ChanThread convertDbPostsToChanThread(String boardName, long threadId, List<PostModel> postModels) {
        if (postModels == null || postModels.size() == 0) {
            return ChanThread.empty();
        }

        List<ChanPost> posts = new ArrayList<>(postModels.size());
        for (PostModel dbPost : postModels) {
            posts.add(dbPost.toPost());
        }

        return new ChanThread(boardName, threadId, posts);
    }

    public static Flowable<Boolean> putThreadFlowable(final ChanThread thread) {
        List<PostModel> posts = convertToPostModels(thread);
        return DatabaseUtils.insert(posts);
    }

    public static boolean putThread(final ChanThread thread) {
        return DatabaseUtils.insertModels(convertToPostModels(thread)) > 0;
    }

    private static List<PostModel> convertToPostModels(ChanThread thread) {
        if (thread == null || thread.getPosts() == null || thread.getPosts().size() == 0) {
            return Collections.emptyList();
        }

        List<PostModel> posts = new ArrayList<>(thread.getPosts().size());
        for (int i = 0; i < thread.getPosts().size(); i++) {
            posts.add(new PostModel(thread.getThreadId(), thread.getPosts().get(i)));
        }

        return posts;
    }

    public static Single<Boolean> removeThread(final long threadId) {
        DatabaseUtils.WhereArg arg = new DatabaseUtils.WhereArg(PostModel.KEY_THREAD_ID + "=?", threadId);
        return DatabaseUtils.remove(new PostModel(), false, arg).first(false);
    }
}

File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/PostTableConnection.kt
package com.emogoth.android.phone.mimi.db

import com.emogoth.android.phone.mimi.db.MimiDatabase.Companion.getInstance
import com.emogoth.android.phone.mimi.db.models.Post
import com.mimireader.chanlib.models.ChanPost
import com.mimireader.chanlib.models.ChanThread
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.functions.Function

object PostTableConnection {
    val LOG_TAG = PostTableConnection::class.java.simpleName

    @JvmStatic
    fun fetchThread(threadId: Long): Single<List<Post>> {
        return getInstance()?.posts()?.getThread(threadId)?.firstOrError()
                ?: Single.just(emptyList())
    }

    @JvmStatic
    fun watchThread(threadId: Long): Flowable<List<Post>> {
        return getInstance()?.posts()?.getThread(threadId) ?: Flowable.just(emptyList())
    }

    @JvmStatic
    fun mapDbPostsToChanThread(boardName: String, threadId: Long): Function<List<Post>, ChanThread> {
        return Function { Posts: List<Post> -> convertDbPostsToChanThread(boardName, threadId, Posts) }
    }

    fun convertDbPostsToChanThread(boardName: String, threadId: Long, posts: List<Post>): ChanThread {
        if (posts.isEmpty()) {
            return ChanThread.empty()
        }
        val p: ArrayList<ChanPost> = ArrayList(posts.size)
        for (dbPost in posts) {
            p.add(dbPost.toPost())
        }
        return ChanThread(boardName, threadId, p)
    }

    @JvmStatic
    fun putThread(thread: ChanThread): Boolean {
        getInstance()?.posts()?.deleteThread(thread.threadId) ?: return false
        getInstance()?.posts()?.insert(convertToPosts(thread)) ?: return false
        return true
    }

    private fun convertToPosts(thread: ChanThread?): List<Post> {
        if (thread == null || thread.posts == null || thread.posts.size == 0) {
            return emptyList()
        }
        val posts: MutableList<Post> = ArrayList(thread.posts.size)
        for (i in thread.posts.indices) {
            posts.add(Post(thread.boardName, thread.threadId, thread.posts[i]))
        }
        return posts
    }

    @JvmStatic
    fun removeThreads(threads: List<Long>): Single<Boolean> {
        return Single.defer {
            getInstance()?.posts()?.deleteThreads(threads) ?: Single.just(false)
            Single.just(true)
        }
    }

    @JvmStatic
    fun removeThread(thread: Long): Single<Boolean> {
        return Single.defer {
            getInstance()?.posts()?.deleteThread(thread) ?: Single.just(false)
            Single.just(true)
        }
    }

    @JvmStatic
    fun fetchThreadIds(): Single<List<Long>> {
        return getInstance()?.posts()?.getFirstPosts()?.firstOrError()
                ?.flatMap { posts: List<Post> ->
                    if (posts.isEmpty()) {
                        return@flatMap Single.just(emptyList<Long>())
                    }
                    val values: ArrayList<Long> = ArrayList(posts.size)
                    for (p in posts) {
                        values.add(p.toPost().no)
                    }
                    Single.just(values)
                } ?: Single.just(emptyList())
    }

    @JvmStatic
    fun removeAllThreads(): Single<Boolean> {
        return Single.defer {
            getInstance()?.posts()?.clear() ?: Single.just(false)
            Single.just(true)
        }
    }
}